moshi/src/test/java/com/squareup/moshi/MoshiTest.java

```diff
@@ -995,6 +995,39 @@
     }
   }
 
+  @Test public void deferredAdapterRecursesInfinitely() throws Exception {
```

Comment:
And I think we cover this in CircularAdaptersTest. I think those tests will fail if you comment out the deferred adapter code.